File: 3.无重复字符的最长子串.java
import java.util.HashSet;

/*
 * @lc app=leetcode.cn id=3 lang=java
 *
 * [3] 无重复字符的最长子串
 */

// @lc code=start
class Solution {
    public int lengthOfLongestSubstring(String s) {
        // System.out.println(s);
        int max = 0;
        int length = s.length();
        HashSet<Character> set = new HashSet<>(length);
        for(int i=0;i<length; i++){
            char c = s.charAt(i);
            // System.out.println(c);
            if(!set.contains(c)){
                set.add(c);
            }else{
                if(max < set.size()){
                    max = set.size();
                }
                set.clear();
            }
        }
        // System.out.println(set);
        return set.size();
    
    }
}
// "pwwkew"
// @lc code=end

File: 3.无重复字符的最长子串.go
/*
 * @lc app=leetcode.cn id=3 lang=golang
 *
 * [3] 无重复字符的最长子串
 */

// @lc code=start
func lengthOfLongestSubstring(s string) int {

	// 用倒排索引解决问题
	var indexArr [255]int

	length := len(strs)
	// 找出长度最长的字符串
	maxStrLen := 0
	for i := 0; i < length; i++ {
		strLength := len(strs[i])
		if strLength > maxStrLen {
			maxStrLen = strLength
		}
	}
	publicStrArr := make([]byte, maxStrLen)
	resStr := make([]byte, maxStrLen)

	t := 0
	res := 0
	// resE := 0
	for i := 0; i < length; i++ {
		cLength := len(strs[i])
		for j := 0; j < cLength; j++ {
			indexArr[strs[i][j]]++
			if indexArr[strs[i][j]] == 3 {
				publicStrArr[t] = strs[i][j]
				t++
				resStr[t] = strs[i][j]
			} else {

				if t > res {
					res = t
					t = 0
				}
			}
		}
	}

	return (string)(publicStrArr)
}

// @lc code=end

